Data fetching guide edited (#201)
withastro/docs: src/pages/en/guides/data-fetching.md

description: Learn how to fetch remote data with Astro using the fetch API.
---

> Astro pages can fetch remote data at build time to help generate your pages.

Reviewer: Mention components can also fetch.

```suggestion
Astro pages and components can fetch remote data at build time to help generate your site.
```

Author: See comment below. I think ONE of these will become `.astro` files, and the other will be Astro components. Might even just come down to which visually strikes me better.

## `fetch()` in Astro

> All [Astro components](/en/core-concepts/astro-components) have access to the global `fetch()` function in their component script to make HTTP requests to APIs. This fetch call will be executed at build time, and the data will be available to the component template for generating dynamic HTML. 

Reviewer: What do you think about using “`.astro` files” as a hand-wavy way to talk about both components and pages?

```suggestion
All `.astro` files have access to the global `fetch()` function in their component script to make HTTP requests to APIs. This fetch call will be executed at build time, and the data will be available to the component template for generating dynamic HTML. 
```

Author: I'm thinking maybe I leave this one here (because I like having an excuse to link to Astro components), and changing the one above into `.astro` files instead of pages and components. (Because pages are just a special case of components anyway, and saying pages and components could inadvertantly give the impression of excluding layouts, eg.)

> All [Astro components](/en/core-concepts/astro-components) have access to the global `fetch()` function in their component script to make HTTP requests to APIs. This fetch call will be executed at build time, and the data will be available to the component template for generating dynamic HTML. 

Reviewer: Given the link to top-level `await` docs, might it also be helpful to link to docs for `fetch` here? For example the equivalent MDN page: https://developer.mozilla.org/en-US/docs/Web/API/fetch

Not sure how best to judge the likelihood that a dev will be familiar with the Fetch API already.

Author: Yeah, I'm undecided as to whether fetch really needs an MDN reference. But, I guess, if it was in there before, and if we do have it for top-level await (which I think is a newer concept, and why I kept it)... all it really does is make the word orange. No big sacrifice for us!

description: Learn how to fetch remote data with Astro using the fetch API.
---

Astro pages can fetch remote data at build time to help generate your pages.

> ## Astro `fetch()`

Reviewer: This is a nit pick but I kept reading this as `Astro.fetch()` which suggests a special API. Maybe something more directly equivalent to the framework components heading could also work?

```suggestion
## `fetch()` on Astro pages
```

description: Learn how to fetch remote data with Astro using the fetch API.
---

`.astro` files can fetch remote data at build time to help generate your pages.

## `fetch()` in Astro

All [Astro components](/en/core-concepts/astro-components) have access to the [global `fetch()` function](https://developer.mozilla.org/en-US/docs/Web/API/fetch) in their component script to make HTTP requests to APIs. This fetch call will be executed at build time, and the data will be available to the component template for generating dynamic HTML. 

> 💡 [**Top-level await**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await) is supported in your Astro component script.

Reviewer 2: This line feels a bit out of place. Since the example below uses it, I don't think you _need_ to call it out here.

Or, if you want to keep a mention, then rephrase to make it clear the reason why you're bringing it up to the user. Something like "top-level await inside of your front matter script makes fetch() even easier to work with" or something like that.

const response = await fetch('https://randomuser.me/api/');
const data = await response.json();
const randomUser = data.results[0]
---
<!-- Data fetched at build can be rendered in HTML -->
<h1>User</h1>
<h2>{randomUser.name.first} {randomUser.name.last}</h2>

<!-- Data fetched at build can be passed to components as props -->
<Contact client:load email={randomUser.email} />
<Location city={randomUser.location.city} />
```

### requestOptions

You can specify any necessary request options for your API fetch call in your Astro page component script.

```astro
---
import MyReactApp from '../components/MyReactApp.jsx';

const headers = new Headers()
headers.append("API_KEY", import.meta.env.API_KEY)

const requestOptions = {
  method: "GET",
  headers: headers,
  redirect: "follow",
  cors:"no-cors"
};

const response = await fetch(
  `https://api.ebird.org/v2/data/obs/CA-PE/recent/notable?detail=full`,
  requestOptions
);
const data = await response.json();
console.log(data);
---
  <MyReactApp dataForApp={data}/>
```

### GraphQL queries

Astro can also use `fetch()` to query a GraphQL server at build time with any valid GraphQL query. 

```astro
---

> import BaseLayout from '../../layouts/BaseLayout.astro';

Reviewer 2: I'd remove the usage of `BaseLayout` from the example, that's the one thing blocking this from being copy-pastable into some user's existing project.

const response = await fetch("https://graphql-weather-api.herokuapp.com", 
  {
    method:'POST',
    headers: {'Content-Type':'application/json'},
    body: JSON.stringify({
      query: `
        query getWeather($name:String!) {
            getCityByName(name: $name){
              name
              country
              weather {
                summary {
                    description
                }
              }
            }
          }
        `,
      variables: {
          name: "Toronto",
      },
    }),
  })

const json = await response.json();
const weather = json.data
---
<BaseLayout title = "GraphQL API fetch in Astro" >
  <h1>Fetching Weather at build time</h2>
  <h2>{weather.getCityByName.name}, {weather.getCityByName.country}</h3>
  <p>Weather: {weather.getCityByName.weather.summary.description}</p>
</BaseLayout>
```

> > 💡 Remember, all data in Astro components is **fetched once, at build time**. For data that can be re-fetched multiple times client side, use a [framework component](/en/core-concepts/framework-components).

Reviewer 2: If you're calling this out, it may make sense to mention the fact that fetch() will be called every time you render a component. So this may mean once during build time, but it may mean on every refresh during dev.

For what it's worth, this is something that we want to fix before v1.0.0 with better caching rules. But for now, that's the reality.

> );

Reviewer: I know this isn’t one of your additions, but does the fetch call here not need an `await`?

Nominating @Princesseuh as a source of truth, if you have a moment.

Author: @RafidMuhymin popped in to the docs channel and suggested await would be consistent with the other examples, and I trust whatever he says! So... what if I just sneak it in there and see if anyone complains?  :)
